dbms/src/Functions/LeastGreatest.h

```diff
@@ -120,5 +121,363 @@ class FunctionVectorizedLeastGreatest : public IFunction
+
+        else
+            res = memcmp(&a_data[0], &b_data[0], std::min(a_size, b_size));
+
+        if constexpr (least)
+        {
+            if (res < 0)
+            {
+                memcpy(&c_data[pre_offset], &a_data[0], a_size);
+                c_offsets[i] = pre_offset + a_size + 1;
+            }
+            else if (res == 0)
+            {
+                size_t size = std::min(a_size, b_size);
+                memcpy(&c_data[pre_offset], &b_data[0], size);
+                c_offsets[i] = pre_offset + size + 1;
+            }
+            else
+            {
+                memcpy(&c_data[pre_offset], &b_data[0], b_size);
+                c_offsets[i] = pre_offset + b_size + 1;
+            }
+        }
+        else
+        {
+            if (res < 0)
+            {
+                memcpy(&c_data[pre_offset], &b_data[0], b_size);
+                c_offsets[i] = pre_offset + b_size + 1;
+            }
+            else if (res == 0)
+            {
+                size_t size = std::max(a_size, b_size);
+                if (a_size > b_size)
+                    memcpy(&c_data[pre_offset], &a_data[0], size);
+                else
+                    memcpy(&c_data[pre_offset], &b_data[0], size);
+
+                c_offsets[i] = pre_offset + size + 1;
+            }
+            else
+            {
+                memcpy(&c_data[pre_offset], &a_data[0], a_size);
+                c_offsets[i] = pre_offset + a_size + 1;
+            }
+        }
+    }
+    // string_string
+    static void process(
+        const TiDB::TiDBCollatorPtr & collator,
+        const ColumnString::Chars_t & a_data,
+        const ColumnString::Offsets & a_offsets,
+        const ColumnString::Chars_t & b_data,
+        const ColumnString::Offsets & b_offsets,
+        ColumnString::Chars_t & c_data,
+        ColumnString::Offsets & c_offsets,
+        size_t i)
+    {
+        size_t a_size = a_offsets[i] - a_offsets[i - 1] - 1;
+        size_t b_size = b_offsets[i] - b_offsets[i - 1] - 1;
+        processImpl(collator, a_size, b_size, &a_data[a_offsets[i - 1]], &b_data[b_offsets[i - 1]], c_data, c_offsets, i);
+    }
+
+    // string_constant
+    static void process(
+        const TiDB::TiDBCollatorPtr & collator,
+        const ColumnString::Chars_t & a_data,
+        const ColumnString::Offsets & a_offsets,
+        const String & b,
+        ColumnString::Chars_t & c_data,
+        ColumnString::Offsets & c_offsets,
+        size_t i)
+    {
+        const auto * b_data = reinterpret_cast<const unsigned char *>(b.data());
+        ColumnString::Offset b_size = b.size();
+        size_t a_size = a_offsets[i] - a_offsets[i - 1] - 1;
+        processImpl(collator, a_size, b_size, &a_data[a_offsets[i - 1]], &b_data[0], c_data, c_offsets, i);
+    }
+
+    // constant_constant
+    static void process(
+        const TiDB::TiDBCollatorPtr & collator,
+        const std::string & a,
+        const std::string & b,
+        std::string & c)
```

Comment:
How about this? We can reduce unnecessary memory allocation and copy.
```suggestion
    static void process(
        const TiDB::TiDBCollatorPtr & collator,
        std::string & a,
        StringRef & b)
```
If the result is a, no memory allocation and copy. Else if the result is b, `a = b.toString()`.

Reply:
Oh. I think we can also use `StringRef` for `a` to avoid any memory allocation.